### ISSUE TITLE
How to change activation function in the Conv Layers of backbone of Yolov11?

### ISSUE THREAD

**Issue by isaadmushtaq** [ultralytics/ultralytics#18595]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey @glenn-jocher, I am experimenting with different activation functions in the Conv layers of YOLOv11. Currently, I am following the approach mentioned in https://github.com/ultralytics/ultralytics/issues/3194#issuecomment-1747787651 and https://github.com/ultralytics/ultralytics/issues/17037#issuecomment-2424358603.

I have changed ultralytics/cfg/models/11/yolo11.yaml as follows,

backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2,None,1,1,'relu']] # 0-P1/2

But I am getting the same results for each activation function. Can you please look into the matter.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @isaadmushtaq, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com) for helpful resources on how to use and customize YOLOv11. You can explore the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples there. These resources may provide insights into model customization, including modifying activation functions.

If this is related to a 🐛 Bug, we kindly request a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/). This helps us understand and debug the issue more effectively. Be sure to include:

- Steps to reproduce the issue  
- Relevant code snippets  
- Error messages or unexpected results  

For custom training or ❓ Questions like yours, please provide detailed information, such as config changes (like your `yolo11.yaml` edits), dataset details, and training logs. Additionally, you might find useful tips in our [Model Training Guide](https://docs.ultralytics.com/guides/model-training-tips/).

To stay connected with the Ultralytics community or to discuss custom changes in-depth, check out any of the following platforms:

- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussions  
- [Discourse](https://community.ultralytics.com) for long-form discussions  
- [Subreddit](https://reddit.com/r/Ultralytics) for open threads with the community  

---

### Upgrade Instructions

Ensure you're working with the latest `ultralytics` package, as bugs and issues may already have been resolved in recent updates. Update your environment by running:

```bash
pip install -U ultralytics
```

Here’s also a quick reminder of the verified environments you can use for YOLO:

- **Notebooks** with free GPUs:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a>  
  <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>  
  <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  

- **Cloud** options like Google Cloud and Amazon's Deep Learning AMIs ([GCP Quickstart](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/), [AWS Quickstart](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))  
- **Docker Images** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>  

---

## Status and CI

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  
Our [Continuous Integration (CI)](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests run automatically across macOS, Windows, and Ubuntu every 24 hours. If the badge here is green, all internal tests are passing as expected ✅.

This is an automated response to assist you promptly. An Ultralytics engineer will follow up with you soon for further support 😊.

**Comment by isaadmushtaq** [comment]
> You have to modify the source code

Can you specify which code you refer to and what modifications I need further?

**Comment by glenn-jocher** [comment]
@isaadmushtaq to experiment with activation functions in YOLOv11, you need to modify the `Conv` module in `ultralytics/nn/modules/conv.py`. Specifically, adjust the `self.act` attribute in the `Conv` class's `__init__` method to use your desired activation function. For more details on this module, refer to its documentation: https://docs.ultralytics.com/reference/nn/modules/conv/. Let me know if further clarification is needed.

**Comment by isaadmushtaq** [comment]
> @isaadmushtaq to experiment with activation functions in YOLOv11, you need to modify the `Conv` module in `ultralytics/nn/modules/conv.py`. Specifically, adjust the `self.act` attribute in the `Conv` class's `__init__` method to use your desired activation function. For more details on this module, refer to its documentation: https://docs.ultralytics.com/reference/nn/modules/conv/. Let me know if further clarification is needed.

Based on, self.act = self.default_act if act is True else act if isinstance(act, nn.Module) else nn.Identity() in `ultralytics/nn/modules/conv.py`. I have two approaches to follow.

#Approach1: (if the act parameter passed to the constructor is explicitly set to True.) 
Change ultralytics/cfg/models/11/yolo11.yaml as follows,

backbone:
  - [-1, 1, Conv, [64, 3, 2,None,1,1,True]] # 0-P1/2

And change `ultralytics/nn/modules/conv.py` as follows,
class Conv(nn.Module):
    default_act = nn.ReLU()

#Approach2:
 Change ultralytics/cfg/models/11/yolo11.yaml as follows,

backbone:
  - [-1, 1, Conv, [64, 3, 2,None,1,1,'ReLU']] # 0-P1/2

And keep 'ultralytics/cfg/models/11/yolo11.yaml' as default.

I am getting same results with each activation fx with approach 1. @glenn-jocher I think I need further clarification.

My train.py file looks like that,
from ultralytics import YOLO

def main():

    model = YOLO('ultralytics/ultralytics/cfg/models/11/custom_yolo11n.yaml')
    
    
    train_results=model.train(
        epochs=20,
        batch=2,
        name="conv5x5 SiLU 20epochs",
        imgsz=640,
        data=r"d:\Amna\Auto_veh\yolo_experiments\vehicle_dataset\Kitti Dataset\Kitti Dataset\kitti.yaml",
        device='cuda'
    )

if __name__ == '__main__':
    main()

**Comment by glenn-jocher** [comment]
The issue likely lies in how the activation function is being interpreted during training. For **Approach 1**, ensure that `default_act` in `ultralytics/nn/modules/conv.py` is correctly set to the desired activation function, and that the `act` parameter in the YAML file is set to `True` to use it. For **Approach 2**, verify that the specified activation (e.g., `'ReLU'`) is being correctly mapped to `nn.ReLU()` inside the `Conv` module. If you're still getting identical results, double-check whether the modified model configuration is being correctly loaded during training. For debugging, print the model architecture using `model.model` after loading to confirm the activation functions. Let me know if further clarification is needed!

**Comment by Y-T-G** [comment]
Did you install your package?

**Comment by zyh321** [comment]
> > The issue likely lies in how the activation function is being interpreted during training. For **Approach 1**, ensure that `default_act` in `ultralytics/nn/modules/conv.py` is correctly set to the desired activation function, and that the `act` parameter in the YAML file is set to `True` to use it. For **Approach 2**, verify that the specified activation (e.g., `'ReLU'`) is being correctly mapped to `nn.ReLU()` inside the `Conv` module. If you're still getting identical results, double-check whether the modified model configuration is being correctly loaded during training. For debugging, print the model architecture using `model.model` after loading to confirm the activation functions. Let me know if further clarification is needed!
> 
> Hey [@glenn-jocher](https://github.com/glenn-jocher), I have debugged both approaches and it's not working.
> 
> # Approach1
> (0): Conv( (conv): Conv2d(3, 16, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False) (bn): BatchNorm2d(16, eps=0.001, momentum=0.03, affine=True, track_running_stats=True) (act): SiLU(inplace=True) )
> 
> # Approach2
> (0): Conv( (conv): Conv2d(3, 16, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False) (bn): BatchNorm2d(16, eps=0.001, momentum=0.03, affine=True, track_running_stats=True) (act): Identity() )
> 
> In approach 2, the first else condition in the 'self.act = self.default_act if act is True else act if isinstance(act, nn.Module) else nn.Identity()' is not being triggered.
> 
> Also, I have tried another approach.
> 
> # Approach3:
> In 'ultralytics/nn/modules/conv.py' I have replaced 'self.act = self.default_act if act is True else act if isinstance(act, nn.Module) else nn.Identity()' with 'self.act = nn.ReLU()'. But it has not effect

Hi， [isaadmushtaq](https://github.com/isaadmushtaq)， 
    Did you resolve this issue? I encountered the same thing.

**Comment by glenn-jocher** [comment]
`act` in `Conv` only supports `True` (uses `Conv.default_act`) or an actual `nn.Module`; passing a string like `'ReLU'` will fall through to `nn.Identity()` by design (see the `Conv` signature/logic in the [Conv module reference](https://docs.ultralytics.com/reference/nn/modules/conv/)). If Approach1/3 “does nothing”, you’re almost certainly editing a different `ultralytics/nn/modules/conv.py` than the one Python is importing (common when the package isn’t installed in editable mode), so please verify paths like this and, if needed, reinstall your clone with `pip install -e .` and restart your Python session/kernel:

```python
import ultralytics, inspect
from ultralytics.nn.modules.conv import Conv

print("ultralytics:", ultralytics.__file__)
print("Conv from:", inspect.getsourcefile(Conv))
print("default_act:", Conv.default_act)
```

If you still can’t get the imported path to match your edited file, please open a new issue (this one is closed) and paste `yolo checks` plus the output above as text so we can see exactly which install is being used.
